src/StoreApi/Utilities/CartController.php

```diff
@@ -360,15 +360,78 @@ private function get_error_message_for_stock_exception_type( $exception_type, $s
 		return __( 'There was an error with an item in your cart.', 'woo-gutenberg-products-block' );
 	}
 
+	/**
+	 * Validate cart and check for errors.
+	 *
+	 * @throws InvalidCartException Exception if invalid data is detected in the cart.
+	 * @throws RouteException Exception if invalid data is detected in the cart.
+	 */
+	public function validate_cart() {
+
+		$cart = $this->get_cart_instance();
+		$this->validate_cart_items();
+		$this->validate_cart_coupons();
+
+		/**
+		 * Fire action to validate cart. Functions hooking into this should return the WP_Error enchanced with custom errors.
+		 *
+		 * @example See docs/examples/validate-cart.md
+		 *
+		 * @param WP_Error $errors  WP_Error object.
+		 * @param WC_Cart  $cart    Cart object.
+		 *
+		 * @return WP_Error         The WP_Error object returned.
+		 */
+		$cart_errors = apply_filters( '__experimental_woocommerce_store_api_cart_errors', new WP_Error(), $cart );
+
+		if ( ! is_wp_error( $cart_errors ) ) {
+
+			throw new RouteException( 'missing_cart_validation_server_error', __( 'There was an error with your cart.', 'woo-gutenberg-products-block' ) );
```

Comment:
Maybe we should switch to a developer facing error here, along these lines: https://github.com/woocommerce/woocommerce-gutenberg-products-block/blob/bbe5b3f9c909139b59c3bea57abd52fb77e1b7ca/src/StoreApi/Schemas/ExtendSchema.php#L311-L319

That way customers won't see strange issues if a developer makes a mistake in a filter, and we can specifically state that consumers of the `__experimental_woocommerce_store_api_cart_errors` filter must return the original `WP_Error` object.

Reply:
In this error case:
- We are dealing with potentially more error messages that were _lost_ in the process, and;
- An invalid cart state could be purchased without issues.

On the other hand, _throwing_ an exception here would disable the store altogether -- no orders can be made either you have errors or don't. 

We might need to consider this as a developer _bug_ that should be handled. I feel like it doesn't worth the risk. Thoughts? 

> we can specifically state that consumers of the __experimental_woocommerce_store_api_cart_errors filter must return the original WP_Error object.

Love this feedback! Regardless of the final message, I thought we could check the original object by using the [spl_object_id](https://www.php.net/manual/en/function.spl-object-id.php) lifetime object id. This way, we can always be sure that no other messages are lost!

Reply:
@xristos3490 Just thought, would it work to create an WP_Error instance and pass it via action hook? That would work as a reference right? Sorry to backtrack.

Reply:
> would it work to create an WP_Error instance and pass it via action hook? That would work as a reference right?

Yes! Calling a simple `do_action` should work by the definition of how PHP handles objects. As long as we use a `WP_Error` instance, every callback in the action's queue will have the same object to handle/add errors in.

This way we can be sure that no one could replace the main iterated error object, but; anyone could _replace/remove/update_ all previous errors.

Although, messing with the error codes can be done with both the `apply_filters` and the `do_action`, approach. 

To sum up, having a single `do_action` that:
- Consolidates all errors,
- Doesn't require anything for the `return` statement, so I can't be broken, and;
- Depends on how PHP works with objects in memory.

Should work perfectly as well!

Reply:
Should be safer, and we'll have inline documentation. Let's go that route. It means we won't have to do any fancy comparison with the WP_Error once the hook has finished running.